atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/PreCommitConditionWithWatches.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.api;
+
+import com.palantir.lock.watch.CommitUpdate;
+
+@FunctionalInterface
+public interface PreCommitConditionWithWatches {
```

Comment:
Please don't add 'WithWatches' etc everywhere in Atlas. Atlas has traditionally suffeered from a pattern where you have a Foo, then you add feature Bar so you have a FooWithBar, and then you add a Baz so you end up with a FooWithBarAndBaz and it all ends up being extremely confusing over time - it lacks abstraction.

Instead, please can we change PreCommitCondition so that it covers this behaviour?

I would suggest that we instead add the new method, have it have a default (empty) implementation, and then have the old method also have a default (empty) implementation. This gives us space to upgrade internal users.

Reply:
It does mean that it's no longer a functional interface - but that's a source and not binary breaking change.